broadcast/src/lib.rs

```diff
@@ -0,0 +1,63 @@
+//! Replication of messages across a network.
+
+use std::future::Future;
+
+use bytes::Bytes;
+use commonware_cryptography::Digest;
+use futures::channel::oneshot;
+use thiserror::Error;
+
+pub mod linked;
+
+/// Errors that can occur when interacting with a stream.
+#[derive(Error, Debug)]
+pub enum Error {
+    #[error("Protobuf decode error")]
+    UnableToDecode,
+    #[error("Duplicate ack")]
+    DuplicateAck,
+    #[error("Conflicting ack")]
+    ConflictingAck,
+    #[error("Unable to create threshold signature")]
+    ThresholdSignature,
+    #[error("Unknown signer")]
+    UnknownSigner,
+}
+
+/// A trait for reliable replication of messages across a network.
+pub trait Broadcaster {
+    type Context;
+
+    /// Broadcast a message to the network.
+    fn broadcast(&mut self, payload: Bytes)
+        -> impl Future<Output = oneshot::Receiver<bool>> + Send;
+}
+
+/// Application is the interface responsible for processing messages received from the network.
+pub trait Application: Send + 'static {
```

Comment:
It'll be important to eventually have a `Supervisor` trait that let's us know which partial signatures we should care about and who the allowed sequencers are.

As discussed, this can be tightly coupled to some sense of `Index` (like in Consensus).